File: ios/Shared/Alarm/IncidentRearm.swift
import Foundation
import UserNotifications

/// Sets the phone's own next ring for an incident the user silenced.
///
/// [RearmRule] is the decision and reads nothing. This finds the five inputs,
/// then puts the ring where the OS can hold it:
///
/// - iOS 26 and later: another AlarmKit alarm under the same
///   `IncidentAlarmScheduler.alarmId(for:)`, set from the Stop intent. The
///   intent runs in the app's own process, so the alarm survives a force-quit,
///   which is the one exit the server repeat cannot reach.
/// - iOS 16 to 25: a Time-Sensitive local notification under the incident id,
///   with the same `alarm.caf` and the `INCIDENT` category. The server repeat
///   lands on the same `apns-collapse-id`, so the two stay one card.
enum IncidentRearm {
    static let category = "INCIDENT"

    /// Silence plus a new ring for the same id. Answers the seconds until that
    /// ring, or nil when nothing was set.
    @discardableResult
    static func rearm(
        incidentId: String,
        now: Date = Date(),
        center: UNUserNotificationCenter? = .current()
    ) async -> Int? {
        let pending = PendingIncidentStore.read(incidentId: incidentId)
        let topic = pending?.topic
        let timers = TopicTimers.read(topic: topic)
        let quietHours = QuietHours.read(from: QuietHours.groupDefaults)

        let allowed = RearmRule.canRearm(
            incidentId: incidentId,
            // Only an incident the alarm path already rang reaches here, and
            // that path runs on open, repeat and reopen alone.
            kind: .repeat,
            criticalOn: TopicTimers.criticalOn(topic: topic),
            ackedLocally: AckedIncidentStore.contains(incidentId: incidentId),
            ringUntil: pending?.ringUntil,
            now: now,
            quietHoursHold: quietHours.holdsRing(
                minuteOfDay: QuietHours.minuteOf(now),
                priority: QuietHours.criticalPriority
            )
        )
        guard allowed else {
            NSLog("CritAlarmAlarm: rearm_skipped incident_id=%@", incidentId)
            return nil
        }

        guard let at = RearmRule.nextRingAt(
            now: now,
            repeatIntervalS: timers?.repeatIntervalS ?? RearmRule.defaultRepeatIntervalS,
            ringUntil: pending?.ringUntil
        ) else {
            NSLog("CritAlarmAlarm: rearm_skipped reason=past_ring_until incident_id=%@", incidentId)
            return nil
        }

        let delay = at.timeIntervalSince(now)
        let seconds = Int(delay.rounded())

        #if canImport(AlarmKit)
        if #available(iOS 26.0, *) {
            let ok = await IncidentAlarmScheduler.schedule(
                incidentId: incidentId,
                topic: pending?.topic ?? "",
                server: pending?.server ?? "",
                title: pending?.title ?? "Crit Alarm",
                sound: pending?.sound,
                delay: delay,
                ringUntil: pending?.ringUntil
            )
            if ok {
                NSLog("CritAlarmAlarm: rearm_set incident_id=%@ in_s=%d", incidentId, seconds)
                return seconds
            }
            // AlarmKit refused, which is what a revoked authorization looks
            // like. Fall through to the notification.
        }
        #endif

        guard let center else { return nil }
        let content = UNMutableNotificationContent()
        content.title = pending?.title ?? "Crit Alarm"
        content.body = "Still ringing. Tap I'm up to end it."
        content.sound = UNNotificationSound(named: UNNotificationSoundName(pending?.sound ?? IncidentAlarmScheduler.soundName))
        content.categoryIdentifier = category
        content.userInfo = ["incident_id": incidentId]
        if #available(iOS 15.0, *) { content.interruptionLevel = .timeSensitive }
        let request = UNNotificationRequest(
            identifier: incidentId,
            content: content,
            trigger: UNTimeIntervalNotificationTrigger(timeInterval: max(delay, 1), repeats: false)
        )
        do {
            try await center.add(request)
            NSLog("CritAlarmAlarm: rearm_notification_set incident_id=%@ in_s=%d", incidentId, seconds)
            return seconds
        } catch {
            NSLog("CritAlarmAlarm: rearm_failed incident_id=%@ error=%@", incidentId, "\(error)")
            return nil
        }
    }

    /// Drops a pending re-arm, whichever of the two holds it. Safe to call
    /// when there is none.
    static func cancel(incidentId: String, center: UNUserNotificationCenter? = .current()) async {
        center?.removePendingNotificationRequests(withIdentifiers: [incidentId])
        #if canImport(AlarmKit)
        if #available(iOS 26.0, *) {
            await IncidentAlarmScheduler.cancel(incidentId: incidentId)
        }
        #endif
        NSLog("CritAlarmAlarm: rearm_cancelled incident_id=%@", incidentId)
    }
}

/// The per-topic timers and the critical switch, as Dart cached them.
///
/// Both come from `GET /v1/topics` and neither is on the push (api.md §5.1),
/// so the re-arm has no other way to learn them without a network call. Dart
/// writes them through shared_preferences, which on iOS is
/// `UserDefaults.standard` with a `flutter.` prefix
/// (`lib/core/notifications/topic_timer_cache.dart`).
struct TopicTimers {
    let repeatIntervalS: Int
    let maxRingS: Int
    let deskTimerS: Int

    static func read(topic: String?, defaults: UserDefaults = .standard) -> TopicTimers? {
        guard let topic, !topic.isEmpty else { return nil }
        return parse(defaults.string(forKey: "flutter.topic_timers.\(topic)"))
    }

    /// The cached value is `repeat_interval_s|max_ring_s|desk_timer_s`.
    static func parse(_ raw: String?) -> TopicTimers? {
        guard let raw else { return nil }
        let parts = raw.split(separator: "|").map { Int($0.trimmingCharacters(in: .whitespaces)) }
        guard parts.count == 3,
              let repeatInterval = parts[0], let maxRing = parts[1], let deskTimer = parts[2],
              repeatInterval > 0, maxRing > 0, deskTimer > 0
        else { return nil }
        return TopicTimers(repeatIntervalS: repeatInterval, maxRingS: maxRing, deskTimerS: deskTimer)
    }

    /// Whether the topic still rings.
    ///
    /// An incident whose topic this device has never listed reads as on: the
    /// push that rang it was a priority-5 alarm push, which the server only
    /// sends for a critical topic, and the server is still repeating anyway.
    static func criticalOn(topic: String?, defaults: UserDefaults = .standard) -> Bool {
        guard let topic, !topic.isEmpty else { return true }
        let key = "flutter.topic_critical.\(topic)"
        guard defaults.object(forKey: key) != nil else { return true }
        return defaults.bool(forKey: key)
    }
}

File: android/app/src/main/kotlin/app/critalarm/alarm/IncidentRearm.kt
package app.critalarm.alarm

import android.content.Context
import android.util.Log
import app.critalarm.push.IncidentPushKind
import app.critalarm.reminders.ReminderReceiver
import app.critalarm.storage.IncidentDeliveryStore
import app.critalarm.storage.NativeConnectionStore
import app.critalarm.storage.TopicTimerStore

/**
 * Sets the phone's own next ring for an incident the user silenced.
 *
 * [RearmRule] is the decision and has no Android in it. This reads the five
 * inputs off disk and hands the answer to [ScheduledAlarmReceiver]. Everything
 * it reads was written by a push or by Dart, so a re-arm only ever re-delivers
 * an incident the server opened.
 */
object IncidentRearm {
    private const val TAG = "CritAlarmAlarm"

    /**
     * Silence plus a new ring at `now + repeat_interval_s`, for the same id.
     * Answers the seconds until that ring, or null when nothing was set.
     */
    fun rearm(
        context: Context,
        incidentId: String,
        title: String? = null,
        body: String? = null,
        nowMillis: Long = System.currentTimeMillis(),
    ): Int? {
        val deliveries = IncidentDeliveryStore(context)
        val topic = deliveries.topicOf(incidentId)
        val timers = topic?.let { TopicTimerStore(context).timersFor(it) }
        val quietHours = QuietHoursWindow.read(context)
        val ringUntil = deliveries.ringUntilMillis(incidentId)

        val allowed = RearmRule.canRearm(
            incidentId = incidentId,
            // Only an incident the alarm path already rang reaches here, and
            // that path runs on open, repeat and reopen alone.
            kind = IncidentPushKind.REPEAT,
            criticalOn = criticalOn(context, topic),
            ackedLocally = deliveries.isAcknowledged(incidentId),
            ringUntilMillis = ringUntil,
            nowMillis = nowMillis,
            quietHoursHold = quietHours.holdsRing(
                minuteOfDay = QuietHoursWindow.minuteOfDay(nowMillis),
                priority = QuietHoursWindow.CRITICAL_PRIORITY,
            ),
        )
        if (!allowed) {
            Log.i(TAG, "rearm_skipped incident_id=$incidentId")
            if (ringUntil != null && nowMillis >= ringUntil) endRingWindow(context, incidentId)
            return null
        }

        val intervalS = timers?.repeatIntervalS ?: RearmRule.DEFAULT_REPEAT_INTERVAL_S
        val at = RearmRule.nextRingAtMillis(
            nowMillis = nowMillis,
            repeatIntervalS = intervalS,
            ringUntilMillis = ringUntil,
        )
        if (at == null) {
            Log.i(TAG, "rearm_skipped reason=past_ring_until incident_id=$incidentId")
            endRingWindow(context, incidentId)
            return null
        }

        val seconds = ((at - nowMillis) / 1000L).toInt()
        val set = ScheduledAlarmReceiver.scheduleAt(
            context = context,
            incidentId = incidentId,
            server = serverOf(context),
            title = title.orEmpty(),
            body = body,
            atMillis = at,
        )
        if (!set) return null
        Log.i(TAG, "rearm_set incident_id=$incidentId in_s=$seconds")
        return seconds
    }

    /**
     * `ring_until` has passed, so this phone will never ring for the incident
     * again. Nobody acked it and nobody closed it, so the only thing to drop
     * is the active flag. Reminders that were waiting for the phone to go
     * quiet go out now, rather than waiting for an ack that is not coming.
     */
    private fun endRingWindow(context: Context, incidentId: String) {
        IncidentDeliveryStore(context).deactivate(incidentId)
        ReminderReceiver.releaseHeld(context)
        Log.i(TAG, "ring_window_over incident_id=$incidentId")
    }

    /** Drops a pending re-arm. Safe to call when there is none. */
    fun cancel(context: Context, incidentId: String) {
        ScheduledAlarmReceiver.cancel(context, incidentId)
        Log.i(TAG, "rearm_cancelled incident_id=$incidentId")
    }

    /**
     * Whether the topic still rings.
     *
     * Dart caches the switch under `topic_critical.<topic>` every time it
     * lists topics. An incident whose topic this device has never listed reads
     * as on: the push that rang it was a priority-5 alarm push, which the
     * server only sends for a critical topic, and the server is still
     * repeating anyway.
     */
    private fun criticalOn(context: Context, topic: String?): Boolean {
        if (topic.isNullOrEmpty()) return true
        val preferences =
            context.getSharedPreferences("FlutterSharedPreferences", Context.MODE_PRIVATE)
        val key = "flutter.topic_critical.$topic"
        if (!preferences.contains(key)) return true
        return preferences.getBoolean(key, true)
    }

    /** One server connection per app in v1, so the re-arm rings against it. */
    private fun serverOf(context: Context): String =
        NativeConnectionStore(context).canonicalServer()?.toString().orEmpty()
}
